### PR TITLE
Implementing the ReviseReplica hook

### LINKED ISSUE #1111: Add interpreter.ReviseReplica hook when create work
**What would you like to be added**:
Recently, I tried to work with ResourceInterpreterWebhook framework which is the new feature in karmada v0.10.0 .
I tried `InterpretReplica` hook, and it work fine for me , But I  found a problem with `ReviseReplica`. 
I cannt find any hook point to call `ReviseReplica`, both in karmada v0.10.0. & v0.10.1 , which like `InterpretReplica ` hook point 
 `ResourceInterpreter.GetReplicas(object) ... ` 

**Why is this needed**:
Now , I interperter my CR's replica and scheduled well, but Karmada Work.workload.manifests replicas cannt be revised so that replica cannot be divided in member cluster. 

Should add  `ReviseReplica` hook point in function `ensureWork` ?

### PR DESCRIPTION
**Pull request by iawia002**
Signed-off-by: Xinzhao Xu <z2d@jifangcheng.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This patch implements the `ReviseReplica` hook, check [resource interpreter webhook proposal](https://github.com/karmada-io/karmada/tree/master/docs/proposals/resource-interpreter-webhook) for more info.

The original code to modify the replica of `Deployment` and `Job` has been changed to use default interpreters, for custom resources, users only need to implement the corresponding webhook and return the patch that updates the replica, take our `workload` resource for example:

```go
func (e *workloadInterpreter) responseWithExploreReviseReplica(workload *workloadv1alpha1.Workload, req interpreter.Request) interpreter.Response {
	wantedWorkload := workload.DeepCopy()
	// update the replica
	wantedWorkload.Spec.Replicas = req.DesiredReplicas
	marshaledBytes, err := json.Marshal(wantedWorkload)
	if err != nil {
		return interpreter.Errored(http.StatusInternalServerError, err)
	}
	// return the patch response
	return interpreter.PatchResponseFromRaw(req.Object.Raw, marshaledBytes)
}
```

I've tested it locally with [the workload example](https://github.com/karmada-io/karmada/tree/master/examples/customresourceinterpreter), everything works as excepted, my test files:

workload:

```yaml
apiVersion: workload.example.io/v1alpha1
kind: Workload
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
```

propagation policy:

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: nginx-workload-propagation
spec:
  resourceSelectors:
    - apiVersion: workload.example.io/v1alpha1
      kind: Workload
      name: nginx
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
    replicaScheduling:
      replicaDivisionPreference: Weighted
      replicaSchedulingType: Divided
      weightPreference:
        staticWeightList:
          - targetCluster:
              clusterNames:
                - member1
            weight: 2
          - targetCluster:
              clusterNames:
                - member2
            weight: 1
```

And in the cluster member1, the workload detail:

```yaml
apiVersion: workload.example.io/v1alpha1
kind: Workload
metadata:
  annotations:
    resourcebinding.karmada.io/name: nginx-workload
    resourcebinding.karmada.io/namespace: default
  creationTimestamp: "2021-12-22T09:16:02Z"
  generation: 1
  labels:
    app: nginx
    propagationpolicy.karmada.io/name: nginx-workload-propagation
    propagationpolicy.karmada.io/namespace: default
    resourcebinding.karmada.io/key: 7f8d449df9
    work.karmada.io/name: nginx-7f8d449df9
    work.karmada.io/namespace: karmada-es-member1
  name: nginx
  namespace: default
  resourceVersion: "5448"
  uid: 8ebe4ee5-d6df-4e62-956c-29e1a3da613b
spec:
  replicas: 2
  template:
    metadata: {}
    spec:
      containers:
      - image: nginx
        name: nginx
        resources: {}
```

In the cluster member2, the workload detail:

```yaml
apiVersion: workload.example.io/v1alpha1
kind: Workload
metadata:
  annotations:
    resourcebinding.karmada.io/name: nginx-workload
    resourcebinding.karmada.io/namespace: default
  creationTimestamp: "2021-12-22T09:16:02Z"
  generation: 1
  labels:
    app: nginx
    propagationpolicy.karmada.io/name: nginx-workload-propagation
    propagationpolicy.karmada.io/namespace: default
    resourcebinding.karmada.io/key: 7f8d449df9
    work.karmada.io/name: nginx-7f8d449df9
    work.karmada.io/namespace: karmada-es-member2
  name: nginx
  namespace: default
  resourceVersion: "5465"
  uid: dd92ed92-8ec9-41a6-a2b2-089173f76085
spec:
  replicas: 1
  template:
    metadata: {}
    spec:
      containers:
      - image: nginx
        name: nginx
        resources: {}
```

**Which issue(s) this PR fixes**:
Fixes #1111

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Implementing the ReviseReplica hook.
```

/cc @RainbowMango @Garrybest @XiShanYongYe-Chang

### CODE REVIEW

**Comment by RainbowMango** [comment]
Pretty good after a quick look at the testing report. I'll review it ASAP.

@GodLoveMinch Is this the feature that you are asking at #1111.

File: pkg/controllers/binding/binding_controller.go
Comment (RainbowMango) [inline pkg/controllers/binding/binding_controller.go:115]: Too much parameters required by `ensureWork` now, maybe you can help review the #1145.

File: pkg/resourceinterpreter/interpreter.go
Comment (RainbowMango) [inline pkg/resourceinterpreter/interpreter.go:105]: Why use `customizedInterpreter.Retain` here? Do you want to share the same logic?

**Review by RainbowMango — COMMENTED** [review]
Review still on-going, post my comments first as I see the base changed.

File: pkg/resourceinterpreter/interpreter.go
Comment (iawia002) [reply pkg/resourceinterpreter/interpreter.go:105]: Actually the logic is the same, I could add a new `customizedInterpreter.ReviseReplica` method but it will be just the same as `customizedInterpreter.Retain`

File: pkg/controllers/binding/binding_controller.go
Comment (iawia002) [reply pkg/controllers/binding/binding_controller.go:115]: Yes but these are all needed, IMHO #1145 just dived one function into several parts, it doesn't solve the too much parameters problem, I think this function will be clearer and more unified after we remove the deprecated RSP API.

**Comment by iawia002** [comment]
> as I see the base changed.

Oh just typo and code format changed, no logic changed.

File: pkg/resourceinterpreter/interpreter.go
Comment (iawia002) [reply pkg/resourceinterpreter/interpreter.go:105]: Maybe we can rename the `customizedInterpreter.Retain` to `Patch`, `ReviseReplica` and `Retain` can share this common method.

File: pkg/resourceinterpreter/utils/utils.go
Comment (RainbowMango) [inline pkg/resourceinterpreter/utils/utils.go:10]: This function just takes `desireReplica` as the common value, it cloud be used to set any fields with `int64` type, right?
I mean the common function is misleading.

By the way, I don't think we should add a `utils` package here, too many `utils` in the same repo is also confusing, we have to use ugly import alias to distinguish them.
Not sure if we need this function yet, if it does, it should be `https://github.com/karmada-io/karmada/blob/master/pkg/util/helper/unstructured.go`.

File: pkg/controllers/binding/common.go
Comment (RainbowMango) [inline pkg/controllers/binding/common.go:113]: The `RSP` has been deprecated and should be cleaned up. I'll open an issue to track it.

File: pkg/resourceinterpreter/interpreter.go
Comment (RainbowMango) [reply pkg/resourceinterpreter/interpreter.go:105]: Not rename, we can abstract a common method shared by `Retain()` and `ReviseReplica()`, for code readability.

File: pkg/resourceinterpreter/utils/utils.go
Comment (iawia002) [reply pkg/resourceinterpreter/utils/utils.go:10]: > it cloud be used to set any fields with int64 type, right?

Yes any fields under `spec`, so should we rename it? Like `ApplySpecInt64Value`, it's weird and there are no others who will use this function.

> it should be https://github.com/karmada-io/karmada/blob/master/pkg/util/helper/unstructured.go.

Sure.

File: pkg/controllers/binding/common.go
Comment (iawia002) [reply pkg/controllers/binding/common.go]: I have not used the `referenceRSP` check logic, this part of the code will be cleaned up separately after the PR merge, #1153

**Comment by iawia002** [comment]
All comments addressed and solved the conflicts, PTAL @RainbowMango

**Comment by RainbowMango** [comment]
got it.
/assign

File: pkg/controllers/binding/binding_controller.go
Comment (RainbowMango) [reply pkg/controllers/binding/binding_controller.go:115]: OK.

**Review by RainbowMango — APPROVED** [review]

**Comment by RainbowMango** [comment]
/lgtm
/hold
have you tested again after you fixed the comments? I'm asking as the E2E test hasn't covered this feature yet.
Feel free to cancel the hold once you confirmed.

@Garrybest 
Just a notice to you, the `Job` related functionality and behavior just work as before.

**Comment by iawia002** [comment]
> have you tested again after you fixed the comments? I'm asking as the E2E test hasn't covered this feature yet.

Will test the deployment case and workload case again.

**Comment by iawia002** [comment]
> > have you tested again after you fixed the comments? I'm asking as the E2E test hasn't covered this feature yet.
> 
> Will test the deployment case and workload case again.

I did find a small problem, the original code panic in case of webhook error because I refer to a nil object, now it has been fixed, others work as expected:

```diff
--- a/pkg/controllers/binding/common.go
+++ b/pkg/controllers/binding/common.go
@@ -113,7 +113,7 @@ func ensureWork(
                        clonedWorkload, err = resourceInterpreter.ReviseReplica(clonedWorkload, desireReplicaInfos[targetCluster.Name])
                        if err != nil {
                                klog.Errorf(“failed to revise replica for %s/%s/%s in cluster %s, err is: %v”,
-                                       clonedWorkload.GetKind(), clonedWorkload.GetNamespace(), clonedWorkload.GetName(), targetCluster.Name, err)
+                                       workload.GetKind(), workload.GetNamespace(), workload.GetName(), targetCluster.Name, err)
                                return err
                        }
```

PATL @RainbowMango

**Comment by RainbowMango** [comment]
Good Job @iawia002

**Pull request merged**
